### DISCUSSION TITLE
How can I make smooth step edges align in height?

### DISCUSSION THREAD

**Discussion by giorgosera** [xyflow/xyflow#2505] [Layouting] [answered]
hey everyone, 

I'm trying to create a workflow builder which allows users to create branch states where based on a condition you can go to one or another branch. 

I'm working with `smoothstep` custom edges and so far I have the following
 
![image](https://user-images.githubusercontent.com/752207/196188522-a3d93ff9-62a9-4ec6-a42c-c1b9c051d197.png)

As you can see the rightmost edge doesn't fold (don't know if this is the right term) at the same height as the other branches. I'm guessing the reason is because the leftmost branch has 2 nodes but the rightmost (and middle one) just one. 

I want to achieve something like this 

![image](https://user-images.githubusercontent.com/752207/196189626-233b5a45-dcad-40f0-869a-64f6432c15c0.png)

For the graph layout I am using Dagre to achieve auto-layout. 

Can someone please point me to an example/documentation/something on how this can be achieved? 

Thanks 🙏

**Comment by moklick** [comment, marked as answer]
This is not built-in and also not very easy to implement. The line runs vertically until it reaches the center of the two nodes. Then it is horizontal. That's why the horizontal lines are not on the same level. You have two options to solve your issue:

1. Use the existing edge logic and create an invisible helper node as seen in this example: https://codesandbox.io/s/bold-haibt-r8lm97?file=/src/Flow.tsx 
2. Implement a custom edge routing that takes node positions into account.

Does this help?

**Reply by giorgosera** [reply]
Yes, that helps indeed! I also thought of adding as many invisible nodes as needed but I hoped there'd be a better way. 

Is there any example of edge routing I can check out? I am not familiar with how this can be done.

**Reply by moklick** [reply]
There is a package for React Flow called [smart-edge](https://github.com/tisoap/react-flow-smart-edge) that you could check out, this is a good article about [orthogonal edge routing](https://medium.com/swlh/routing-orthogonal-diagram-connectors-in-javascript-191dc2c5ff70) or you could take a look at [elkjs](https://github.com/kieler/elkjs) which has node and edge routing algorithms.

**Reply by giorgosera** [reply]
Thank you so much for the help and the awesome library you've created! :) Keep rocking man!

**Reply by giorgosera** [reply]
@moklick would the above be possible in the paid version of react-flow? I can see that there is  an auto-layout feature in the pro version.

**Reply by giorgosera** [reply]
I managed to create the layout exactly as I want it with elkjs and I have a visual example [here](https://rtsys.informatik.uni-kiel.de/elklive/json.html?compressedContent=N4KABGCWAmBcYCIBOB7FAXBAacYA2AhgJ4oCu6A8gA7qQoB2AzvMGLhAOQF4DmKSkdAAsAth3gdCRAKZJp0DjgidpeANYA6aJDkBjWg3FgOAEQoB1AHKL2x3Q0YxZAWRTRVFJO6QbG6JATo0jxERhyWFCYAogDKAPoAgpYmcVEmAOKxNsrG-Dwa0rp4kFSM0gXqGlKy8lXEspD0+X4BQSFhAJKWACpRAEoJAMLdHQBqUdnKHHkFRSVlFZrVctAa9G7SAAqEutIi0vToGgBGmgBmkAAe8gnFPPT7h2EAQgkAMkmDaRy4AL5KYF0QkgeGgcnosAA2rYIKwYPAEPQAIzYMAAdxgwngAGYAAxYMBCaSQHhCdA4-H4AjHVTMMCQ1hBS7k4zIji-AC6YH+MLAcLgiHoACZURjoFiwHiCUSSWSKQTCDS8HSGWAmSyOML2VyeTlYVABYjsaLMUJ5YTiaSWVKqUqVYzpMyJPRsdruQCcvyEfQACwm8VmyWUmVW82K2nwVXq50+t26vVewUAVn9EptIblQYV1Ij9IdTtZSbjHuUicRADZU4H05bMzbw8rI-mNfRy8XeWX6AB2KvmjPWykN+1qx0trtx3AcgHyHjSFW2MvSFEExhkJC7FWClFgKdqghIWfoTeIkU790Lg0I6QildrjeRwWn3fofeH48uhBn+P6+GIaTG29SHXOcHyNT9n1faQj1A31wPPT1Lz-P1AOA98nwJF8Dyg98Uy-EtFxTFD73pR9wIwyDoJIis4O-PlEIQaRKyIkCqNgndyKwyjIUFHs8IvX8GJ7ZicLIvdOPfXidXw+jpAADlRVcgOI7jqPYsS3xgyTuUnEBfiAA) in their editor.

However, when I use the same setup in react-flow the rightmost edge is still bending halfway (see below). 

![image](https://user-images.githubusercontent.com/752207/196449719-91f26745-43a0-44ed-be00-5066d953869f.png)

I am guessing that somehow in my custom edge code I'll have to remove the `getSmoothStepPath` I'm using now and use the one calculated by elkjs. 

Is that correct? How do I skip react flow's edge routing and use the one calculated by Elkjs?

**Reply by moklick** [reply]
The pro examples are advanced examples. Everything is built with the React Flow library so you can build all those things without a subscription. Please see my first answer. It's still the same case here you just moved the node on the very right to the top, now you would need to add an invisible edge below.

**Reply by giorgosera** [reply]
Thanks @moklick. It's clear about the pro version.

Regarding the invisible node I decided not to do that because it would be hard to generalize and handle cases like nested branches etc. The complexity adds up.

I decided to take your advice regarding elkjs and their edge routing. I'm getting layouted edges sections as provided by Elkjs and generating an SVG path using d3-shape. 

```javascript
import { line } from 'd3-shape';

const toSvgPath = (edge: ElkExtendedEdge) => {
  const { sections } = edge;
  return (sections ?? []).map(section => {
    const data = [section.startPoint, ...(section.bendPoints || []), section.endPoint];
    return line<ElkPoint>().x(d => d.x).y(d => d.y)(data);
  });
}
```

Then I inject that `path` in the edge data.

```javascript
const layoutedEdges = edges.map(e => {
  const elkEdge = newGraph?.edges?.find(({ id }) => id === e.id);
  if (!elkEdge) return e;

  const path = toSvgPath(elkEdge);
  e.data = e.data ?? {};
  e.data.path = path;
  return e;
});
```

And finally in my custom edge code I use the SVG path generated by ELK and d3.

```javascript
export default function StepEdge({
  id,
  sourceX,
  sourceY,
  targetX,
  targetY,
  sourcePosition,
  targetPosition,
  style = {},
  markerEnd,
  label,
  labelShowBg,
  labelStyle,
  data,
}: EdgeOptions) {

  const { labelX, labelY, path } = data;

  return (
    <>
      <path
        id={id}
        style={style}
        className="react-flow__edge-path"
        d={data?.path}
        markerEnd={markerEnd}
      />
      {label && <EdgeText
        x={targetX}
        y={targetY - 30}
        label={label} 
        labelStyle={labelStyle}
        labelShowBg={labelShowBg}
        labelBgPadding={[2, 4]}
        labelBgBorderRadius={2}
      />}
      <foreignObject
        width={foreignObjectSize}
        height={foreignObjectSize}
        x={labelX - foreignObjectSize / 2}
        y={labelY - foreignObjectSize / 2}
        className="step-edge__foreignobject"
        requiredExtensions="http://www.w3.org/1999/xhtml"
      >
        <div className='step_edge__body'>
          <button className="step_edge" onClick={(event) => onEdgeClick(event, id)}>
            +
          </button>
        </div>
      </foreignObject>
    </>
  );
}
```

I'm getting closer but I'm still getting a path which is slightly off the rest of the graph (see below). It's the correct shape but with an offset. Do you have any clue why this might be happening?

<img width="731" alt="image" src="https://user-images.githubusercontent.com/752207/196535243-e1d4ca7c-43da-4851-9ef4-ce9a99d72927.png">

**Reply by giorgosera** [reply]
I found the issue. I was not using the correct width and height for my nodes during layouting. I have to wait for the nodes to be initialized with these dimensions and then apply any layouting.

**Reply by djibomar** [reply]
Hey @giorgosera I'm facing a similar issue where the nodes are not properly centered. Did you have to use partitioning to get this to work? Thanks 
I found your thread here https://github.com/kieler/elkjs/issues/192

**Reply by Kritica1** [reply]
I'm having the same issue, if you figure it out, please share

**Comment by dongtc** [comment]
```
const [edgePath] = getSmoothStepPath({
      sourceX,
      sourceY,
      sourcePosition,
      targetX,
      targetY,
      centerY: sourceY + 50,
      targetPosition,
    });
```

I found the same issue, and have a solution by using `centerY: sourceY + 50,`
